pilot/pkg/xds/ads.go

```diff
@@ -257,17 +275,48 @@ func (s *DiscoveryServer) StreamAggregatedResources(stream discovery.AggregatedD
 			}
 
 		case pushEv := <-con.pushChannel:
-			// TODO: possible race condition: if a config change happens while the envoy
-			// was getting the initial config, between LDS and RDS, the push will miss the
-			// monitored 'routes'. Same for CDS/EDS interval. It is very tricky to handle
-			// due to the protocol - but the periodic push recovers from it.
-			err := s.pushConnection(con, pushEv)
+			s.pushConnection(con, pushEv)
 			pushEv.done()
-			if err != nil {
-				return nil
+			return nil
+		}
+	}
+}
+
+// Manage the ack if EnableFlowControl is set.
+// Benchmark looks good with this initial take. :-)
+func (s *DiscoveryServer) manageAck(con *Connection, typeUrl string) error {
+	if con.proxy.WatchedResources == nil || con.proxy.WatchedResources[typeUrl] == nil {
+		return nil
+	}
+
+	ctx := context.TODO()
+
+	if features.EnableFlowControl {
+		// Retrieve the nonce sent and the nonce acked
+		con.proxy.RLock()
+		acked := con.proxy.WatchedResources[typeUrl].NonceAcked
+		sent := con.proxy.WatchedResources[typeUrl].NonceSent
+		con.proxy.RUnlock()
+
+		// Try to acquire the semaphore. If the semaphore is in use,
+		// the last time the noncees were compared, they were not equivalent.
+		// In this case, compare the noncees again and release the semaphore
+		// if they are equivalent. This may be doable with a RWlock.
+		acquire := con.sem[typeUrl].TryAcquire(1)
+		if acquire == false {
+			if acked == sent {
+				con.sem[typeUrl].Release(1)
 			}
 		}
+
+		// Always acquire the semaphore. This logic isn't quite right, as
+		// we should probably compare the nonces prior to acquisition. This
+		// will result in the next unmatched nonce executing the logic above.
+		if err := con.sem[typeUrl].Acquire(ctx, 1); err != nil {
```

Comment:
I don't see how this works - maybe I am missing something. I would expect that `TryAcquire` on L305 succeeds, thus making the semaphore filled. Then we block here forever trying to acquire again?

Reply:
next time through with the same resource type, tryAcquire is false, releasing the semaphore if the nonce's match. In the case that the lock is held,315 returns with a nil. When spamming requests through Istio, this works well. I am going to attempt refactoring the code so it is a little more intiuitive.

In other cases the lock is acquire the lock via 315. As the  comments state (this is an early version), if the nonce is not equivalent, we should acquire the lock. The code doesn't represent this as I haven't had a chance to test.

Thanks for the early review. I was mostly interested in your take as if this blocks all pushes and slows down pilot - as it is in the push path and part of the 100 limit.

Thank you John.

Cheers,
-steve